README.md
CSS Media Query
===============

[![Build Status](https://travis-ci.org/ericf/css-mediaquery.png?branch=master)](https://travis-ci.org/ericf/css-mediaquery)
[![Dependency Status](https://gemnasium.com/ericf/css-mediaquery.png)](https://gemnasium.com/ericf/css-mediaquery)
[![npm Version](https://badge.fury.io/js/css-mediaquery.png)](https://npmjs.org/package/css-mediaquery)

Parses and determines if a given CSS Media Query matches a set of values via
JavaScript.

Installation
------------

Install via npm

```
npm install css-mediaquery
```

Usage
-----

This package has two exports: `parse()`, and `match()` which can parse CSS Media
Queries and determine if a media query matches a given set of values.

### Matching

The `match()` method lets you compare a media query expression with a JavaScript
object and determine if a media query matches a given set of values.

```javascript
var mediaQuery = require('css-mediaquery');

var isMatch = mediaQuery.match('screen and (min-width: 40em)', {
    type : 'screen',
    width: '1024px'
});

console.log(isMatch); // => true
```

### Parsing

Existing CSS Parsers don't do a great job at parsing the details of media
queries. That's where `css-mediaquery` shines. You can parse a media query
expression and get an AST back by using the `parse()` method.

```javascript

var mediaQuery = require('css-mediaquery');

var ast = mediaQuery.parse('screen and (min-width: 48em)');
```

The `ast` variable will have the following payload:

```javascript
[ { inverse: false, type: 'screen', expressions: [
    { modifier: 'min', feature: 'width', value: '48em' }
] } ]
```

This package was written with care to following the W3C Recommendations for
[CSS3 Media Queries][w3c-mq] and [CSS3 Values and Units][w3c-vu]. It supports
all of the [Media Features][w3c-mq-features] and will properly convert values to
a common unit before comparing them.


The values specified to check a media query string against should be thought of
as if they are the current state of a device/browser. A `type` value _must_ be
specified, and it can _not_ be `"all"`.


[w3c-mq]: http://www.w3.org/TR/css3-mediaqueries/
[w3c-mq-features]: http://www.w3.org/TR/css3-mediaqueries/#media1
[w3c-vu]: http://www.w3.org/TR/css3-values/


License
-------

This software is free to use under the Yahoo! Inc. BSD license.
See the [LICENSE file][] for license text and copyright information.


[LICENSE file]: https://github.com/ericf/css-mediaquery/blob/master/LICENSE


HISTORY.md
CSS Media Query Change History
==============================

NEXT
----

* Made `parse()` sticter; it will now thorw a `SyntaxError` if a media query is
  invalid. This makes it behave more like `JSON.parse()`.


0.1.2 (2014-01-21)
------------------

* Fixed issue with parsing media queries that do not have expressions. ([#5][])


[#5]: https://github.com/ericf/css-mediaquery/issues/5


0.1.1 (2014-01-08)
------------------

* Added docs to README for `parse()`.


0.1.0 (2014-01-08)
------------------

* Initial public release.


test/unit-tests.js
/* global describe, it */

'use strict';

var expect     = require('chai').expect,
    mediaQuery = require('../');

describe('mediaQuery.parse()', function () {
    it('should parse media queries without expressions', function () {
        expect(mediaQuery.parse('screen')).to.eql([
            {
                inverse    : false,
                type       : 'screen',
                expressions: []
            }
        ]);

        expect(mediaQuery.parse('not screen')).to.eql([
            {
                inverse    : true,
                type       : 'screen',
                expressions: []
            }
        ]);
    });

    it('should parse common retina media query list', function () {
        var parsed = mediaQuery.parse(
            'only screen and (-webkit-min-device-pixel-ratio: 2),\n' +
            'only screen and (   min--moz-device-pixel-ratio: 2),\n' +
            'only screen and (     -o-min-device-pixel-ratio: 2/1),\n' +
            'only screen and (        min-device-pixel-ratio: 2),\n' +
            'only screen and (                min-resolution: 192dpi),\n' +
            'only screen and (                min-resolution: 2dppx)'
        );

        expect(parsed).to.be.an.array;
        expect(parsed).to.have.length(6);
        expect(parsed[0].expressions[0].feature).to.equal('-webkit-min-device-pixel-ratio');
        expect(parsed[1].expressions[0].modifier).to.equal('min');
    });

    it('should throw a SyntaxError when a media query is invalid', function () {
        function parse(query) {
            return function () { mediaQuery.parse(query); };
        }

        expect(parse('some crap')).to.throw(SyntaxError);
        expect(parse('48em')).to.throw(SyntaxError);
        expect(parse('screen and crap')).to.throw(SyntaxError);
        expect(parse('screen and (48em)')).to.throw(SyntaxError);
        expect(parse('screen and (foo:)')).to.throw(SyntaxError);
        expect(parse('()')).to.throw(SyntaxError);
        expect(parse('(foo) (bar)')).to.throw(SyntaxError);
        expect(parse('(foo:) and (bar)')).to.throw(SyntaxError);

        //These should fail
        expect(parse('foo')).to.throw(SyntaxError);
        expect(parse('foo and (max-width: 48em)')).to.throw(SyntaxError);
    });
});

describe('mediaQuery.match()', function () {
    describe('Equality Check', function () {
        it('Orientation: should return true for a correct match (===)', function () {
            expect(mediaQuery.match(
                '(orientation: portrait)', {orientation: 'portrait'}
            )).to.be.true;
        });

        it('Orientation: should return false for an incorrect match (===)', function () {
            expect(mediaQuery.match(
                '(orientation: landscape)', {orientation: 'portrait'}
            )).to.be.false;
        });

        it('Scan: should return true for a correct match (===)', function () {
            expect(mediaQuery.match(
                '(scan: progressive)', {scan: 'progressive'}
            )).to.be.true;
        });

        it('Scan: should return false for an incorrect match (===)', function () {
            expect(mediaQuery.match(
                '(scan: progressive)', {scan: 'interlace'}
            )).to.be.false;
        });

        it('Width: should return true for a correct match', function () {
            expect(mediaQuery.match(
                '(width: 800px)', {width: 800}
            )).to.be.true;
        });

        it('Width: should return false for an incorrect match', function () {
            expect(mediaQuery.match(
                '(width: 800px)', {width: 900}
            )).to.be.false;
        });
    });

    describe('Length Check', function () {
        describe('Width', function () {
            it('should return true for a width higher than a min-width', function () {
                expect(mediaQuery.match(
                    '(min-width: 48em)', {width: '80em'}
                )).to.be.true;
            });

            it('should return false for a width lower than a min-width', function () {
                expect(mediaQuery.match(
                    '(min-width: 48em)', {width: '20em'}
                )).to.be.false;
            });

            it('should return false when no width value is specified', function () {
                expect(mediaQuery.match(
                    '(min-width: 48em)', {resolution: 72}
                )).to.be.false;
            });
        });

        describe('Different Units', function () {
            it('should work with ems', function () {
                expect(mediaQuery.match(
                    '(min-width: 500px)', {width: '48em'}
                )).to.be.true;
            });

            it('should work with rems', function () {
                expect(mediaQuery.match(
                    '(min-width: 500px)', {width: '48rem'}
                )).to.be.true;
            });

            it('should work with cm', function () {
                expect(mediaQuery.match(
                    '(max-height: 1000px)', {height: '20cm'}
                )).to.be.true;
            });

            it('should work with mm', function () {
                expect(mediaQuery.match(
                    '(max-height: 1000px)', {height: '200mm'}
                )).to.be.true;
            });

            it('should work with inch', function () {
                expect(mediaQuery.match(
                    '(max-height: 1000px)', {height: '20in'}
                )).to.be.false;
            });

            it('should work with pt', function () {
                expect(mediaQuery.match(
                    '(max-height: 1000px)', {height: '850pt'}
                )).to.be.false;
            });

            it('should work with pc', function () {
                expect(mediaQuery.match(
                    '(max-height: 1000px)', {height: '60pc'}
                )).to.be.true;
            });
        });

    });

    describe('Resolution Check', function () {
        it('should return true for a resolution match', function () {
            expect(mediaQuery.match(
                '(resolution: 50dpi)', {resolution: 50}
            )).to.be.true;
        });

        it('should return true for a resolution higher than a min-resolution', function () {
            expect(mediaQuery.match(
                '(min-resolution: 50dpi)', {resolution: 72}
            )).to.be.true;
        });

        it('should return false for a resolution higher than a max-resolution', function () {
            expect(mediaQuery.match(
                '(max-resolution: 72dpi)', {resolution: 300}
            )).to.be.false;
        });

        it('should return false if resolution isnt passed in', function () {
            expect(mediaQuery.match(
                '(min-resolution: 72dpi)', {width: 300}
            )).to.be.false;
        });

        it('should convert units properly', function () {
            expect(mediaQuery.match(
                '(min-resolution: 72dpi)', {resolution: '75dpcm'}
            )).to.be.false;

            expect(mediaQuery.match(
                '(resolution: 192dpi)', {resolution: '2dppx'}
            )).to.be.true;
        });
    });

    describe('Aspect Ratio Check', function () {
        it('should return true for an aspect-ratio higher than a min-aspect-ratio', function () {
            expect(mediaQuery.match(
                '(min-aspect-ratio: 4/3)', {'aspect-ratio': '16 / 9'}
            )).to.be.true;
        });

        it('should return false for an aspect-ratio higher than a max-aspect-ratio', function () {
            expect(mediaQuery.match(
                '(max-aspect-ratio: 4/3)', {'aspect-ratio': '16/9'}
            )).to.be.false;
        });

        it('should return false if aspect-ratio isnt passed in', function () {
            expect(mediaQuery.match(
                '(max-aspect-ratio: 72dpi)', {width: 300}
            )).to.be.false;
        });

        it('should work numbers', function () {
            expect(mediaQuery.match(
                '(min-aspect-ratio: 2560/1440)', {'aspect-ratio': 4 / 3}
            )).to.be.false;
        });
    });

    describe('Grid/Color/Color-Index/Monochrome', function () {
        it('should return true for a correct match', function () {
            expect(mediaQuery.match(
                '(grid)', {grid: 1}
            )).to.be.true;

            expect(mediaQuery.match(
                '(color)', {color: 1}
            )).to.be.true;

            expect(mediaQuery.match(
                '(color-index: 3)', {'color-index': 3}
            )).to.be.true;

            expect(mediaQuery.match(
                '(monochrome)', {monochrome: 1}
            )).to.be.true;
        });

        it('should return false for an incorrect match', function () {
            expect(mediaQuery.match(
                '(grid)', {grid: 0}
            )).to.be.false;

            expect(mediaQuery.match(
                '(color)', {color: 0}
            )).to.be.false;

            expect(mediaQuery.match(
                '(color-index: 3)', {'color-index': 2}
            )).to.be.false;

            expect(mediaQuery.match(
                '(monochrome)', {monochrome: 0}
            )).to.be.false;

            expect(mediaQuery.match(
                '(monochrome)', {monochrome: 'foo'}
            )).to.be.false;
        });
    });

    describe('Type', function () {
        it('should return true for a correct match', function () {
            expect(mediaQuery.match(
                'screen', {type: 'screen'}
            )).to.be.true;
        });

        it('should return false for an incorrect match', function () {
            expect(mediaQuery.match(
                'screen and (color:1)', {
                    type : 'tv',
                    color: 1
                }
            )).to.be.false;
        });

        it('should return false for a media query without a type when type is specified in the value object', function () {
            expect(mediaQuery.match(
                '(min-width: 500px)', {type: 'screen'}
            )).to.be.false;
        });

        it('should return true for a media query without a type when type is not specified in the value object', function () {
            expect(mediaQuery.match(
                '(min-width: 500px)', {width: 700}
            )).to.be.true;
        });
    });

    describe('Not', function () {
        it('should return false when theres a match on a `not` query', function () {
            expect(mediaQuery.match(
                'not screen and (color)', {
                    type : 'screen',
                    color: 1
                }
            )).to.be.false;
        });

        it('should not disrupt an OR query', function () {
            expect(mediaQuery.match(
                'not screen and (color), screen and (min-height: 48em)', {
                    type  : 'screen',
                    height: 1000
                }
            )).to.be.true;
        });

        it('should return false for when type === all', function () {
            expect(mediaQuery.match(
                'not all and (min-width: 48em)', {
                    type : 'all',
                    width: 1000
                }
            )).to.be.false;
        });

        it('should return true for inverted value', function () {
            expect(mediaQuery.match(
                'not screen and (min-width: 48em)', {width: '24em'}
            )).to.be.true;
        });
    });

    describe('mediaQuery.match() Integration Tests', function () {
        describe('Real World Use Cases (mostly AND)', function () {
            it('should return true because of width and type match', function () {
                expect(mediaQuery.match(
                    'screen and (min-width: 767px)', {
                        type : 'screen',
                        width: 980
                    }
                )).to.be.true;
            });

            it('should return true because of width is within bounds', function () {
                expect(mediaQuery.match(
                    'screen and (min-width: 767px) and (max-width: 979px)', {
                        type : 'screen',
                        width: 800
                    }
                )).to.be.true;
            });

            it('should return false because width is out of bounds', function () {
                expect(mediaQuery.match(
                    'screen and (min-width: 767px) and (max-width: 979px)', {
                        type : 'screen',
                        width: 980
                    }
                )).to.be.false;
            });

            it('should return false since monochrome is not specified', function () {
                expect(mediaQuery.match(
                    'screen and (monochrome)', {width: 980}
                )).to.be.false;
            });

            it('should return true since color > 0', function () {
                expect(mediaQuery.match(
                    'screen and (color)', {
                        type : 'screen',
                        color: 1
                    }
                )).to.be.true;
            });

            it('should return false since color = 0', function () {
                expect(mediaQuery.match(
                    'screen and (color)', {
                        type : 'screen',
                        color: 0
                    }
                )).to.be.false;
            });
        });

        describe('Grouped Media Queries (OR)', function () {
            it('should return true because of color', function () {
                expect(mediaQuery.match(
                    'screen and (min-width: 767px), screen and (color)', {
                        type : 'screen',
                        color: 1
                    }
                )).to.be.true;
            });

            it('should return true because of width and type', function () {
                expect(mediaQuery.match(
                    'screen and (max-width: 1200px), handheld and (monochrome)', {
                        type : 'screen',
                        width: 1100
                    }
                )).to.be.true;
            });

            it('should return false because of monochrome mis-match', function () {
                expect(mediaQuery.match(
                    'screen and (max-width: 1200px), handheld and (monochrome)', {
                        type      : 'screen',
                        monochrome: 0
                    }
                )).to.be.false;
            });
        });
    });
});



index.js
/*
Copyright (c) 2014, Yahoo! Inc. All rights reserved.
Copyrights licensed under the New BSD License.
See the accompanying LICENSE file for terms.
*/

(function (root, factory) {
    'use strict';

    // UMD
    /* istanbul ignore next */
    if (typeof define === 'function' && define.amd) {
        define(factory);
    } else if (typeof exports === 'object') {
        module.exports = factory();
    } else if (typeof YUI !== 'undefined' && YUI.add) {
        YUI.add('css-mediaquery', factory, '0.1.2', {es: true});
    } else {
        root.mediaQuery = factory();
    }

})(typeof global !== 'undefined' ? global : this, function () {

// -----------------------------------------------------------------------------

'use strict';

var exports = {
    match: matchQuery,
    parse: parseQuery
};

// -----------------------------------------------------------------------------

var RE_MEDIA_QUERY     = /^(?:(only|not)?\s*([_a-z][_a-z0-9-]*)|(\([^\)]+\)))(?:\s*and\s*(.*))?$/i,
    RE_MQ_EXPRESSION   = /^\(\s*([_a-z-][_a-z0-9-]*)\s*(?:\:\s*([^\)]+))?\s*\)$/,
    RE_MQ_FEATURE      = /^(?:(min|max)-)?(.+)/,
    RE_LENGTH_UNIT     = /(em|rem|px|cm|mm|in|pt|pc)?\s*$/,
    RE_RESOLUTION_UNIT = /(dpi|dpcm|dppx)?\s*$/;

function matchQuery(mediaQuery, values) {
    return parseQuery(mediaQuery).some(function (query) {
        var inverse = query.inverse;

        // Either the parsed or specified `type` is "all", or the types must be
        // equal for a match.
        var typeMatch = query.type === 'all' || values.type === query.type;

        // Quit early when `type` doesn't match, but take "not" into account.
        if ((typeMatch && inverse) || !(typeMatch || inverse)) {
            return false;
        }

        var expressionsMatch = query.expressions.every(function (expression) {
            var feature  = expression.feature,
                modifier = expression.modifier,
                expValue = expression.value,
                value    = values[feature];

            // Missing or falsy values don't match.
            if (!value) { return false; }

            switch (feature) {
                case 'orientation':
                case 'scan':
                    return value.toLowerCase() === expValue.toLowerCase();

                case 'width':
                case 'height':
                case 'device-width':
                case 'device-height':
                    expValue = toPx(expValue);
                    value    = toPx(value);
                    break;

                case 'resolution':
                    expValue = toDpi(expValue);
                    value    = toDpi(value);
                    break;

                case 'aspect-ratio':
                case 'device-aspect-ratio':
                case /* Deprecated */ 'device-pixel-ratio':
                    expValue = toDecimal(expValue);
                    value    = toDecimal(value);
                    break;

                case 'grid':
                case 'color':
                case 'color-index':
                case 'monochrome':
                    expValue = parseInt(expValue, 10) || 1;
                    value    = parseInt(value, 10) || 0;
                    break;
            }

            switch (modifier) {
                case 'min': return value >= expValue;
                case 'max': return value <= expValue;
                default   : return value === expValue;
            }
        });

        return (expressionsMatch && !inverse) || (!expressionsMatch && inverse);
    });
}

function parseQuery(mediaQuery) {
    return mediaQuery.split(',').map(function (query) {
        query = query.trim();

        var captures = query.match(RE_MEDIA_QUERY);

        // Media Query must be valid.
        if (!captures) {
            throw new SyntaxError('Invalid CSS media query: "' + query + '"');
        }

        var modifier    = captures[1],
            type        = captures[2],
            expressions = ((captures[3] || '') + (captures[4] || '')).trim(),
            parsed      = {};

        parsed.inverse = !!modifier && modifier.toLowerCase() === 'not';
        parsed.type    = type ? type.toLowerCase() : 'all';

        // Check for media query expressions.
        if (!expressions) {
            parsed.expressions = [];
            return parsed;
        }

        // Split expressions into a list.
        expressions = expressions.match(/\([^\)]+\)/g);

        // Media Query must be valid.
        if (!expressions) {
            throw new SyntaxError('Invalid CSS media query: "' + query + '"');
        }

        parsed.expressions = expressions.map(function (expression) {
            var captures = expression.match(RE_MQ_EXPRESSION);

            // Media Query must be valid.
            if (!captures) {
                throw new SyntaxError('Invalid CSS media query: "' + query + '"');
            }

            var feature = captures[1].toLowerCase().match(RE_MQ_FEATURE);

            return {
                modifier: feature[1],
                feature : feature[2],
                value   : captures[2]
            };
        });

        return parsed;
    });
}

// -- Utilities ----------------------------------------------------------------

function toDecimal(ratio) {
    var decimal = Number(ratio),
        numbers;

    if (!decimal) {
        numbers = ratio.match(/^(\d+)\s*\/\s*(\d+)$/);
        decimal = numbers[1] / numbers[2];
    }

    return decimal;
}

function toDpi(resolution) {
    var value = parseFloat(resolution),
        units = String(resolution).match(RE_RESOLUTION_UNIT)[1];

    switch (units) {
        case 'dpcm': return value / 2.54;
        case 'dppx': return value * 96;
        default    : return value;
    }
}

function toPx(length) {
    var value = parseFloat(length),
        units = String(length).match(RE_LENGTH_UNIT)[1];

    switch (units) {
        case 'em' : return value * 16;
        case 'rem': return value * 16;
        case 'cm' : return value * 96 / 2.54;
        case 'mm' : return value * 96 / 2.54 / 10;
        case 'in' : return value * 96;
        case 'pt' : return value * 72;
        case 'pc' : return value * 72 / 12;
        default   : return value;
    }
}

return exports;

});
